lib/elixir/lib/registry.ex

```diff
@@ -922,6 +922,11 @@ defmodule Registry do
         ok
 
       {:error, {:already_registered, ^self}} = error ->
+        # This is an error for the :unique case, so there are two entries.
+        # :ets.delete_object deletes both since pid_ets is a :duplicate_bag so
+        # add one back.
+        true = :ets.delete_object(pid_ets, {self, key, key_ets})
```

Comment:
Unfortunately this is a race condition. Because we have a key registered under `self()`, if this process happens to crash right after deleting here due to a link, its key will never be removed. I believe the proper fix is to do what @michalmuskala suggested and replace [this line](https://github.com/elixir-lang/elixir/blob/77e36f5b98d33cc2d1cbb02fbda572576cb84bce/lib/elixir/lib/registry.ex#L914) by `insert_new` and return `{:error, {:already_registered, self()}}` straight away.

Reply:
I tried to use `insert_new` to create a cleanup callback that would only delete when `insert_new` returned true. I ran into a lot of failing tests and didn't get a chance to jump back to it. 

@fhunleth, did you have failures like:

~~~bash
 23) test unique with 8 partitions allows process unregistering (RegistryTest)
     test/elixir/registry_test.exs:184
     Assertion with == failed
     code:  assert Registry.keys(registry, self()) |> Enum.sort() == ["hello", "world"]
     left:  ["hello"]
     right: ["hello", "world"]
     stacktrace:
       test/elixir/registry_test.exs:189: (test)

.............

 24) test duplicate with 1 partition compares using matches (RegistryTest)
     test/elixir/registry_test.exs:300
     Assertion with == failed
     code:  assert Registry.keys(registry, self()) |> Enum.sort() == [1, 1.0]
     left:  [1.0]
     right: [1, 1.0]
     stacktrace:
       test/elixir/registry_test.exs:303: (test)
~~~

That is just the last two because I thought putting them all in here would be a terrible plan.

Reply:
Yes, I ran into that. There's some more conversation on the linked issue. The current plan of attack is to add a unique number to the tuple that gets put in ETS so that it can be deleted on fail (José's original idea). I got waylaid by work or I'd be continuing down that path. It needs a bunch of updates to other code to make it work, though.